rllib/models/tf/tf_action_dist.py

```diff
@@ -64,32 +64,31 @@ def deterministic_sample(self) -> TensorType:
 
     @override(ActionDistribution)
     def logp(self, x: TensorType) -> TensorType:
-        return -tf.nn.sparse_softmax_cross_entropy_with_logits(
-            logits=self.inputs, labels=tf.cast(x, tf.int32)
-        )
+        probs = tf.nn.softmax(self.inputs)
+        probs = tf.clip_by_value(probs, tf.keras.backend.epsilon(), 1.0)
+        logp = tf.math.log(probs)
+        return tf.gather(logp, x, batch_dims=1)
 
     @override(ActionDistribution)
     def entropy(self) -> TensorType:
-        a0 = self.inputs - tf.reduce_max(self.inputs, axis=1, keepdims=True)
-        ea0 = tf.exp(a0)
-        z0 = tf.reduce_sum(ea0, axis=1, keepdims=True)
-        p0 = ea0 / z0
-        return tf.reduce_sum(p0 * (tf.math.log(z0) - a0), axis=1)
+        probs = tf.nn.softmax(self.inputs)
+        probs = tf.clip_by_value(probs, tf.keras.backend.epsilon(), 1.0)
+        return tf.reduce_sum(probs * tf.math.log(probs), axis=1)
 
     @override(ActionDistribution)
     def kl(self, other: ActionDistribution) -> TensorType:
-        a0 = self.inputs - tf.reduce_max(self.inputs, axis=1, keepdims=True)
-        a1 = other.inputs - tf.reduce_max(other.inputs, axis=1, keepdims=True)
-        ea0 = tf.exp(a0)
-        ea1 = tf.exp(a1)
-        z0 = tf.reduce_sum(ea0, axis=1, keepdims=True)
-        z1 = tf.reduce_sum(ea1, axis=1, keepdims=True)
-        p0 = ea0 / z0
-        return tf.reduce_sum(p0 * (a0 - tf.math.log(z0) - a1 + tf.math.log(z1)), axis=1)
+        probs = tf.nn.softmax(self.inputs)
+        probs = tf.clip_by_value(probs, tf.keras.backend.epsilon(), 1.0)
+        probsOther = tf.nn.softmax(other.inputs)
+        probsOther = tf.clip_by_value(probsOther, tf.keras.backend.epsilon(), 1.0)
+        return tf.keras.losses.kl_divergence(probs, probsOther)
 
     @override(TFActionDistribution)
     def _build_sample_op(self) -> TensorType:
-        return tf.squeeze(tf.random.categorical(self.inputs, 1), axis=1)
+        probs = tf.nn.softmax(self.inputs)
+        probs = tf.clip_by_value(probs, tf.keras.backend.epsilon(), 1.0)
+        logp = tf.math.log(probs)
+        return tf.squeeze(tf.random.categorical(logp, 1), axis=1)
```

Comment:
I think this will break action masking and is my only concern about merging this right now.

Some actions must have zero prob sometimes, if they are masked by the env.

@HJasperson

Reply:
Can we just use 0.0 here instead of tf.keras.backend.epsilon() for the lower bound?

Reply:
But probs are not allowed to be zero anyways if we put them into tf.math.log(probs) afterwards (where they will be undefined).

Reply:
Ah, sorry, true, log probs would fail. Just felt like tf.keras.backend.epsilon (10e-7) is still too large.

Reply:
tf.keras.backend.epsilon is what they use in the tf source code, so that's why I used it. I ran some tests, though, and it looks like tf,math.log should return real numbers down to about 1e-37. I can replace tf.keras.backend.epsilon with 1e-37 if y'all prefer.

Reply:
I think tensorflow should accept tensor initialization with -np.inf, no?

Reply:
tf.constant([-np.inf]) works on my end! So I think that would be the cleanest way to go. Sorry for the back and forth here and thanks for sticking with us!

Reply:
Are we still talking about `tf.clip_by_value(probs, tf.keras.backend.epsilon(), 1.0)`? The reason you need this line is because tf.math.log returns -inf if you get too close to 0. The smallest (positive) number you can input and still return a real value is 1e-37. -np.inf does not address this at all.

Reply:
Absolutely. Sorry I should have been more clear.
`tf.math.log(probs)` should be -np.inf where self.inputs is zero imo.

Reply:
So we'll have to "manually" set the logprobs to -np.inf in these instances.

Reply:
Can we do `tf.float32.min` instead, then? This is the number that our e.g. EpsilonExploration uses to figure out whether it should pick an action (randomly) or not?